### PR TITLE
br: add compaction after data resolver and test 

### PR DESCRIPTION
**Pull request by fengou1**
### What is changed and how it works?
1. add compaction after data resolved
2. add more test
3. remove some redundancy code

Issue Number: Close #xxx

What's Changed:

```commit-message
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 
xTZzcAIe3g

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
Please add a release note.
If you don't think this PR needs a release note then fill it with None.
If this PR will be picked to release branch, then a release note is probably required.
```

### CODE REVIEW

File: components/raftstore/src/store/peer.rs [hunk 1]
```diff
@@ -611,7 +611,7 @@ pub struct SnapshotRecoveryWaitApplySyncer {
 }
 
 impl SnapshotRecoveryWaitApplySyncer {
-    pub fn new(region_id: u64, sender: SyncSender<u64>) -> Self {
+    pub fn new(region_id: u64, sender: StdSender<u64>) -> Self {
```
Comment (3pointer) [inline components/raftstore/src/store/peer.rs]: why change it?

File: Cargo.toml [hunk 2]
```diff
@@ -208,10 +208,7 @@ cmake = { git = "https://github.com/rust-lang/cmake-rs" }
 
 [target.'cfg(target_os = "linux")'.dependencies]
 procinfo = { git = "https://github.com/tikv/procinfo-rs", rev = "6599eb9dca74229b2c1fcc44118bef7eff127128" }
-# When you modify TiKV cooperatively with kvproto, this will be useful to submit the PR to TiKV and the PR to
-# kvproto at the same time.
-# After the PR to kvproto is merged, remember to comment this out and run `cargo update -p kvproto`.
-
+# When you modify TiKV cooperatively with kvproto, this will be useful to submit the PR to TiKV and the PR to# kvproto at the same time.# After the PR to kvproto is merged, remember to comment this out and run `cargo update -p kvproto`.
```
Comment (3pointer) [inline Cargo.toml]: shouldn't change it🤔

File: components/snap_recovery/src/services.rs [hunk 3]
```diff
@@ -417,6 +459,7 @@ impl<ER: RaftEngine> RecoverData for RecoveryService<ER> {
                 Ok((resp, WriteFlags::default()))
             });
             sink.send_all(&mut s).await?;
+            compact(db.clone()).expect("compact kvdb failure");
```
Comment (3pointer) [inline components/snap_recovery/src/services.rs]:
```suggestion
            compact(db.clone())?
```

File: components/snap_recovery/src/services.rs [hunk 4]
```diff
@@ -257,6 +262,38 @@ impl<ER: RaftEngine> RecoveryService<ER> {
...
+    for cf_name in engine.cf_names() {
+        let cf = cf_name.to_owned().clone();
+        let kv_db = engine.clone();
+        let h = Builder::new()
+            .name(format!("compact-{}", cf))
+            .spawn_wrapper(move || {
+                info!("recovery starts manual compact"; "cf" => cf.clone());
+                tikv_alloc::add_thread_memory_accessor();
+                let db = kv_db.as_inner();
+                let handle = get_cf_handle(db, cf.as_str()).unwrap();
```
Comment (3pointer) [inline components/snap_recovery/src/services.rs:146]: should handle this error.

File: components/snap_recovery/src/services.rs [hunk 5]
```diff
@@ -257,6 +262,38 @@ impl<ER: RaftEngine> RecoveryService<ER> {
...
+                let mut compact_opts = CompactOptions::new();
+                compact_opts.set_max_subcompactions(64);
+                compact_opts.set_exclusive_manual_compaction(false);
+                compact_opts.set_bottommost_level_compaction(DBBottommostLevelCompaction::Skip);
+                db.compact_range_cf_opt(handle, &compact_opts, None, None);
+                tikv_alloc::remove_thread_memory_accessor();
+                info!("recovery finishes manual compact"; "cf" => cf);
+            })
+            .expect("failed to spawn compaction thread");
```
Comment (3pointer) [inline components/snap_recovery/src/services.rs:156]:
```suggestion
            ?
```

File: components/raftstore/src/store/peer.rs (hunk 1 above)
Comment (fengou1) [reply components/raftstore/src/store/peer.rs]: the thought is async channel may be better, however, according to the test, it looks does not matter, so to keep identical with wait apply PR, change back to SyncSender.

File: Cargo.toml (hunk 2 above)
Comment (fengou1) [reply Cargo.toml]: it looks like feature branch need rebase

File: components/snap_recovery/src/services.rs (hunk 4 above)
Comment (fengou1) [reply components/snap_recovery/src/services.rs:146]: fixed

File: components/snap_recovery/src/services.rs (hunk 3 above)
Comment (fengou1) [reply components/snap_recovery/src/services.rs]: it looks we need error handling here.

File: components/snap_recovery/src/services.rs (hunk 5 above)
Comment (fengou1) [reply components/snap_recovery/src/services.rs:156]: the trait `From<std::io::Error>` is not implemented for `kvproto::recoverdatapb::Error`

File: components/snap_recovery/src/services.rs [hunk 6]
```diff
@@ -171,73 +94,10 @@ impl<ER: RaftEngine> RecoveryService<ER> {
         ];
 
         let tmp_opts: Vec<_> = opts.iter().map(|(k, v)| (k.as_str(), v.as_str())).collect();
-        self.engines
-            .kv
-            .set_options_cf(cf_name, tmp_opts.as_slice())
-            .unwrap();
+        engine.set_options_cf(cf_name, tmp_opts.as_slice()).unwrap();
```
Comment (3pointer) [inline components/snap_recovery/src/services.rs:96]:
```suggestion
        engine.set_options_cf(cf_name, tmp_opts.as_slice())?;
```

File: components/snap_recovery/src/region_meta_collector.rs [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+            if suffix != keys::REGION_STATE_SUFFIX {
+                return Ok(true);
+            }
+            regions.push(id);
+            Ok(true)
+        }));
+
+        for region_id in regions {
+            let region_state_key = keys::region_state_key(region_id);
+            let region_state = box_try!(
```
Comment (3pointer) [inline components/snap_recovery/src/region_meta_collector.rs]: can we use pattern match here?
like
```
match (Some(raft_state), Some(region_state), Some(apply_state)) => {
     if region_state.get_state() == PeerState::Tombstone {
                continue;
    }
   ....
}

**Comment by 3pointer** [comment]
rest LGTM

**Review by 3pointer — APPROVED** [review]

**Pull request merged**